cmake/Sanitizers.cmake

```diff
@@ -0,0 +1,94 @@
+# Both of these files are listed in the compile commands JSON, however, with
+# the current tidy target implementation, the sanitizer cannot know about them.
+# This results in bad/corrupted PCH errors and missing compile commands (errors)
+# for source files included in unity builds.
+if(USE_SANITIZE_TARGETS AND USE_CLANG_TIDY)
+    set(ALLOW_BUILD_OPTIMIZATIONS FALSE)
+else()
+    set(ALLOW_BUILD_OPTIMIZATIONS TRUE)
+endif()
+
+cmake_dependent_option(
+    USE_CLANG_TIDY
+    "Allow the codebase to use the clang-tidy sanitizer"
+    ${ALLOW_BUILD_OPTIMIZATIONS} # Default such that we don't clobber PlasmaTargets.
+    "CLANG_TIDY_EXE"
+    OFF
+)
+
+
+if(USE_CLANG_TIDY)
+    # Don't set CMAKE_CXX_CLANG_TIDY. CMake+Ninja+MSVC+clang-tidy are broken in this mode, see:
+    # https://gitlab.kitware.com/cmake/cmake/-/issues/20512
+    # https://bugs.llvm.org/show_bug.cgi?id=45356
+    # Use the tidy target instead.
+    if(USE_SANITIZE_TARGETS)
+        set(_TIDY_REPLACEMENTS_DIR "${CMAKE_BINARY_DIR}/clang-tidy-replacements")
+        add_custom_target(tidy_init
+            COMMENT "Preparing to tidy..."
+            COMMAND ${CMAKE_COMMAND} -E remove_directory "${_TIDY_REPLACEMENTS_DIR}"
+            COMMAND ${CMAKE_COMMAND} -E make_directory "${_TIDY_REPLACEMENTS_DIR}"
+        )
+        add_custom_target(tidy DEPENDS tidy_init)
+        if(CLANG_APPLY_EXE)
+            add_custom_target(fix
+                DEPENDS tidy
+                COMMENT "Applying fixes..."
+                COMMAND ${CLANG_APPLY_EXE} "${_TIDY_REPLACEMENTS_DIR}"
+                WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}"
+            )
+        endif()
+
+        # If we were on CMake 3.20, we could do this per-target...
+        set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
+    endif()
+endif()
+
+function(plasma_sanitize_target TARGET)
+    if(USE_CLANG_TIDY)
+        # CMake only natively supports CXX_CLANG_TIDY with Unix Makefiles and Ninja generator as of
+        # CMake 3.19. But most everyone is using Visual Studio, so we'll reset the code analysis
+        # type to clang-tidy. This is a huge win given how terrible MS's code analysis is.
+        if(CMAKE_GENERATOR MATCHES "Visual Studio")
+            set_target_properties(
+                ${TARGET} PROPERTIES
+                VS_GLOBAL_ClangTidyToolExe "${CLANG_TIDY_EXE}"
+                VS_GLOBAL_EnableClangTidyCodeAnalysis TRUE
+                VS_GLOBAL_EnableMicrosoftCodeAnalysis FALSE
+            )
+        endif()
+
+        if(USE_SANITIZE_TARGETS)
+            get_target_property(_TARGET_SOURCES ${TARGET} SOURCES)
+            get_target_property(_TARGET_SOURCE_DIR ${TARGET} SOURCE_DIR)
+            foreach(_SOURCE_FILE IN LISTS _TARGET_SOURCES)
+                # Prevent CMake from stoopidly feeding Python files and bitmaps to clang-tidy.
+                # Yes, CMake actually does try to do that...
+                if(_SOURCE_FILE MATCHES [[\.c$|\.cpp$|\.cxx$|\.cc$]])
```

Comment:
No headers?

Reply:
We can only tidy files listed in the compile commands JSON. I think clang-tidy smart enough to discard headers, knowing that they'll be checked on include. It isn't smart enough to filter out python files, resource files, or bitmaps though.

Reply:
Very odd, but it raises an interesting question:  Why are those files listed as target sources to begin with?  Can't they be added to the source group without being a target source?

Reply:
Unfortunately not. Omitting them from the target prevents them from being listed in the IDE at all. Also, you want *.rc files to be compiled into the target by MSVC.